File: rotate_image.cxx
class Solution {
public:
    
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        for(int j=0;j<n/2;++j) {
            int loops = n - 1 - 2*j;
            for(int i=0;i<loops;++i) {
                int temp = matrix[i+j][n-1-j];
                matrix[i+j][n-1-j] = matrix[j][i+j];
                matrix[j][i+j] = matrix[n-1-i-j][j];
                matrix[n-1-i-j][j] = matrix[n-1-j][n-1-i-j];
                matrix[n-1-j][n-1-i-j] = temp;
            }
        }
    }
    
    /*void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        for(int i=0;i<n;++i)
            for(int j=0;j<i;++j)
                std::swap(matrix[i][j],matrix[j][i]);
        for(int i=0;i<n;++i)
            std::reverse(matrix[i].begin(),matrix[i].end());
    }*/    
    
};
 
/*
    n=3
    1(0,0)   0,n-1 -> n-1,n-1 -> n-1,0 -> 0,0
    2(0,1)   1,n-1 -> n-1,n-2 -> n-2,0 -> 0,1
    
    n=4
    5(0,0)   0,n-1 -> n-1,n-1 -> n-1,0 -> 0,0
    1(0,1)   1,n-1 -> n-1,n-2 -> n-2,0 -> 0,1
    9(0,2)   2,n-1 -> n-1,n-3 -> n-3,0 -> 0,2
    4(1,1)   1,n-2 -> n-2,n-2 -> n-2,1 -> 1,1       
*/

File: rotate_image.py
#!/usr/bin/env python3

from typing import List

class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        n=len(matrix)
        #transpose
        for i in range(n):
            for j in range(i,n):
                #print(f'[{i},{j}]')
                matrix[i][j],matrix[j][i] = matrix[j][i],matrix[i][j]

        #reverse
        for i in range(n):
            for j in range(n//2):
                print(f'[{i},{j}]')
                matrix[i][j],matrix[i][n-j-1] = matrix[i][n-j-1],matrix[i][j]





if __name__ == "__main__":
    matrix = [[1,2,3],[4,5,6],[7,8,9]]
    print(matrix)
    #print(rotate(matrix))
    Solution().rotate(matrix)

